### PR TITLE
Feat/experiment_manager

### PR DESCRIPTION
**Pull request by ihsanKisi**
## 📥 Pull Request Description

This pull request implements the experiment manager feature. This feature was requested to be able to filter experiments based on their properties.

## 👀 Affected Areas

Experiments

## 📝 Checklist

Please make sure you've completed the following tasks before submitting this pull request:

- [x] Pre-commit hooks were executed
- [x] Changes have been reviewed by at least one other developer
- [x] Tests have been added or updated to cover the changes (only necessary if the changes affect the executable code)
- [x] All tests ran successfully
- [x] All merge conflicts are resolved
- [ ] Documentation has been updated to reflect the changes
- [ ] Any necessary migrations have been run

## 📌 Related Issues

[Feature]: We have to manage the experiments in the dashboard. #2

## 🔗 Links

Please provide any relevant links (e.g. documentation, external resources) that support your changes.

## 📷 Screenshots

Without filtering (10 experiments):
image

With filtering by experiment_type:
image

### CODE REVIEW

**Review by aiakide — CHANGES_REQUESTED** [review]
Please merge the current  `develop` branch. Please pay attention to which files you commit. Not everything should be added to the repo.

**Review by Jankeyy — APPROVED** [review]

File: nicemldashboard/State/State.py [hunk 1]
```diff
@@ -4,22 +4,29 @@
...
+    """
 
 
 class EventManager:
     """
     Manages event dictionaries and provides methods to retrieve and manage them.
     """
 
-    _observable_dicts: dict[str, ObservableDict]
+    _observable_dicts: dict[StateKeys, ObservableDict]
```
Comment (aiakide) [inline nicemldashboard/State/State.py:29]: Please use a more appropriate variable name than `_observable_dicts`. This is the data that is stored in our AppState.

File: nicemldashboard/State/State.py [hunk 2]
```diff
@@ -4,22 +4,29 @@
...
+_EVENT_MNGR_ATTR_NAME = "_nicemldashboard"
+
+
+class StateKeys(ABCMeta, EnumMeta):
+    """
+    Provides an abstract class for StateKey management
+    """
 
 
 class EventManager:
```
Comment (aiakide) [inline nicemldashboard/State/State.py:24]: Please use a more appropriate class name. This is where our app data is stored. I qould suggest something like `State`, or `AppState`.

File: nicemldashboard/State/State.py [hunk 3]
```diff
@@ -77,9 +84,23 @@ def init_event_manager(event_manager: EventManager):
         logger.debug("Event Manager cannot be initialized in a background task")
```
Comment (aiakide) [inline nicemldashboard/State/State.py:84]: Is this a critical message? Then please use logger.warn. debug is probably rarely used as a log level.

File: nicemldashboard/State/State.py
Comment (aiakide) [inline nicemldashboard/State/State.py:1]: Please rename the package. In Python, it is common to name packages in lowercase letters. 
The file name should be the same as the name you have chosen for the class. Currently still `EventManager`.

File: nicemldashboard/basecomponents/sidebar.py [hunk 4]
```diff
@@ -13,6 +13,14 @@
 from typing import Optional, List
 
 from nicegui import ui
+from nicegui.observables import ObservableDict
+
+
+from nicemldashboard.State.State import (
+    _get_event_manager,
```
Comment (aiakide) [inline nicemldashboard/basecomponents/sidebar.py:20]: Variable and method names begin with an underscore if they are to be declared as private. However, as you are importing the method here, it is no longer private. Please remove the leading underscore

File: nicemldashboard/basecomponents/sidebar.py [hunk 5]
```diff
@@ -36,14 +44,44 @@ def sidebar(experiment_types: Optional[List[ExperimentType]] = None):
         side_bar_toggle = SidebarToggleButton(left_drawer=left_drawer)
 
         for experiment_type in experiment_types:
+            button_callback = lambda e, et=experiment_type: (
```
Comment (aiakide) [inline nicemldashboard/basecomponents/sidebar.py:47]: That looks a bit hacky. Maybe you can implement a new button in `buttons.py` (`SidebarButton`) that has a `StateManager` and overrides the `on_click` method. The button also gets an `ExperimentType` when it is initialised. The for loop then initialises a button for each `ExperimentType`.

File: nicemldashboard/basecomponents/sidebar.py [hunk 6]
```diff
@@ -36,14 +44,44 @@ def sidebar(experiment_types: Optional[List[ExperimentType]] = None):
...
+                color="transparent", icon=experiment_type.icon, on_click=button_callback
+            ).props("flat").classes("exp-type-btn").bind_text_from(
                 experiment_type,
                 "prefix",
                 backward=lambda x: ("" if not side_bar_toggle.is_expanded() else x),
             ):
                 ui.tooltip(experiment_type.name)
+
+
+def on_stage_change(observable_dict: ObservableDict, value: str):
```
Comment (aiakide) [inline nicemldashboard/basecomponents/sidebar.py:64]: See the comment above. I don't think we need the method any more

File: nicemldashboard/basecomponents/sidebar.py [hunk 7]
```diff
@@ -36,14 +44,44 @@ def sidebar(experiment_types: Optional[List[ExperimentType]] = None):
...
+    """
+    Updates the Events.on_experiment_change key with the value given.
+    :param observable_dict:
+    :param value:
+    :return:
+    """
+    observable_dict[ExperimentEvents.ON_EXPERIMENT_PREFIX_CHANGE] = value
+
+
+def get_enum(experiment_type: ExperimentType):
```
Comment (aiakide) [inline nicemldashboard/basecomponents/sidebar.py:74]: See the comment above. I don't think we need the method any more

File: nicemldashboard/basecomponents/sidebar.py (hunk 5 above)
Comment (aiakide) [reply nicemldashboard/basecomponents/sidebar.py:47]: We will probably get more button classes. Please make a package of the file `buttons.py` and create a separate module / file for each button.

File: nicemldashboard/experiment/experiment_manager.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    def __init__(self, experiments: List[Experiment]):
+        """
+        Initializes an ExperimentManager with the provided list of experiments.
+        """
+        self.experiments = experiments
+
+    def filter_by(self, **filters) -> List[Experiment]:
+        """
+        Filters the experiment list after filtering with the provided filters
```
Comment (aiakide) [inline nicemldashboard/experiment/experiment_manager.py:24]: Please use google-style docstrings

File: nicemldashboard/experiment/experiment_manager.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        :param filters:
+        :return:
+        """
+        filtered_experiments = self.experiments.copy()
+        for key, value in filters.items():
```
Comment (aiakide) [inline nicemldashboard/experiment/experiment_manager.py:29]: Can you find better names than `key` and `value`?

File: nicemldashboard/experiment/experiment_manager.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            try:
+                if isinstance(value, dict):
```
Comment (aiakide) [inline nicemldashboard/experiment/experiment_manager.py:31]: I find this confusing, I would expect `filter` to be used as follows. 

experiment_manager.filter(experiment_id="abs233", git_version": "0.0.1: asdasd, 0.0.2: dasdad").

Therefore value is never a dict. Probably only `str`, `int`, `float`, `bool`. Other filters can be added later.

File: nicemldashboard/experiment/experiment_manager.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                            for item in value.items()
+                        )
+                    ]
+                else:
+                    filtered_experiments = [
+                        exp
+                        for exp in filtered_experiments
+                        if getattr(exp, key, None) == value
+                    ]
+            except TypeError as e:
```
Comment (aiakide) [inline nicemldashboard/experiment/experiment_manager.py:46]: Please create your own exception that can inherit from TypeError. Maybe something like `ExperimentFilterError`.

File: nicemldashboard/pages/home.py [hunk 12]
```diff
@@ -10,21 +10,36 @@
...
 @ui.page("/")
 def home():
     """
     Define the layout of the home page.
     """
+
     ui.add_scss("nicemldashboard/assets/style.scss")
+    _instance = EventManager()
+    init_event_manager(_instance)
+    exp_dict = _instance.get_dict(ExperimentStateKeys.EXPERIMENT_DICT)
```
Comment (aiakide) [inline nicemldashboard/pages/home.py:38]: Rename `exp_dict` into something like `experiment_state` oder `experiment_data`

File: nicemldashboard/pages/home.py [hunk 13]
```diff
@@ -10,21 +10,36 @@
...
+    exp_dict.on_change(_experiment_runs_table.refresh)
```
Comment (aiakide) [inline nicemldashboard/pages/home.py:39]: I don't think experiment_runs_table should be declared as private with a leading underscore.

File: nicemldashboard/pages/home.py [hunk 14]
```diff
@@ -33,4 +48,14 @@ def home():
                 with ui.row():
                     ui.input(label="Experiment run", placeholder="Search for run")
                 ui.separator()
-                experiment_runs_table(experiments=experiments)
+                _experiment_runs_table(experiment_manager, exp_dict)
+
+
+@ui.refreshable
+def _experiment_runs_table(
```
Comment (aiakide) [inline nicemldashboard/pages/home.py:55]: Please move this method in `table.py`

File: pyproject.toml [hunk 15]
```diff
@@ -27,6 +27,7 @@ pytest = "^8.1.1"
 select = ["E","F","D100", "D101","D102", "D103","D105","D106","D107","PT","PL"]
 fixable = ["E","F","D","PT","PL"]
 exclude = ["./tests/**"]
+ignore = ["E731"]
```
Comment (aiakide) [inline pyproject.toml:30]: This rule makes sense. Lamdas are generally more difficult to read and should be avoided if possible.

File: tests/unit/nicemldashboard/experiment/test_experiment_manager.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+def experiment_manager(experiments):
+    return ExperimentManager(experiments)
+
+
+def test_filter_by_name(experiment_manager, experiments):
+    name_to_test = experiments[0].name
+    filtered = experiment_manager.filter_by(name=name_to_test)
+    assert len(filtered) == 1
+
+    non_filtered_experiments = [exp for exp in experiments if exp.name != name_to_test]
```
Comment (aiakide) [inline tests/unit/nicemldashboard/experiment/test_experiment_manager.py:21]: Why is the filtered experiment no longer included in the `experiments`? The filter method should only return the filtered experiment(s), but should not be removed from the list of experiments.

File: tests/unit/nicemldashboard/experiment/test_experiment_manager.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+def test_filter_by_id(experiment_manager, experiments):
+    id_to_test = experiments[0].experiment_id
+    filtered = experiment_manager.filter_by(experiment_id=id_to_test)
+    assert len(filtered) == 1
+
+    non_filtered_experiments = [
+        exp for exp in experiments if exp.experiment_id != id_to_test
+    ]
+    assert all(exp.experiment_id != id_to_test for exp in non_filtered_experiments)
```
Comment (aiakide) [inline tests/unit/nicemldashboard/experiment/test_experiment_manager.py:33]: Same here

File: tests/unit/nicemldashboard/experiment/test_experiment_manager.py [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def test_filter_by_experiment_type(experiment_manager, experiments):
+    experiment_type_to_test = experiments[0].experiment_type
+    filtered = experiment_manager.filter_by(experiment_type=experiment_type_to_test)
+    assert all(exp.experiment_type == experiment_type_to_test for exp in filtered)
+
+    experiments_not_filtered = [exp for exp in experiments if exp not in filtered]
+    assert all(
```
Comment (aiakide) [inline tests/unit/nicemldashboard/experiment/test_experiment_manager.py:42]: Same here

File: tests/unit/nicemldashboard/experiment/test_experiment_manager.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    assert len(filtered) == 0
+
+
+def test_filter_by_multiple_fields_name_and_id(experiment_manager, experiments):
+    first_experiment = experiments[0]
+    filtered = experiment_manager.filter_by(
+        name=first_experiment.name, experiment_id=first_experiment.experiment_id
+    )
+    assert len(filtered) == 1
+    experiments_not_filtered = [exp for exp in experiments if exp not in filtered]
```
Comment (aiakide) [inline tests/unit/nicemldashboard/experiment/test_experiment_manager.py:59]: Same here

File: tests/unit/nicemldashboard/experiment/test_experiment_manager.py [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    filtered = experiment_manager.filter_by(
+        experiment_type=first_experiment.experiment_type,
+        git_version=first_experiment.git_version,
+    )
+    assert all(
+        exp.experiment_type == first_experiment.experiment_type for exp in filtered
+    )
+    assert all(exp.git_version == first_experiment.git_version for exp in filtered)
+
+    experiments_not_filtered = [exp for exp in experiments if exp not in filtered]
```
Comment (aiakide) [inline tests/unit/nicemldashboard/experiment/test_experiment_manager.py:80]: Same here

**Review by aiakide — CHANGES_REQUESTED** [review]
Please have a look at the comments. We can talk about it again.

File: nicemldashboard/experiment/experiment_manager.py
Comment (aiakide) [inline nicemldashboard/experiment/experiment_manager.py:1]: Please remove underscore from filename

**Review by aiakide — APPROVED** [review]

**Pull request merged**